### ISSUE TITLE
Add CUDA callback to Python API

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#38671]
As discussed in [this forum thread](https://discuss.pytorch.org/t/mpi-cuda-stream/80702/4). CUDA stream/event callback can be a useful feature. It might be helpful to add it to the Python API. 

Besides, some recent discussion for TensorPipe RPC backend also considered using CUDA callback to sync streams and handle errors. Hence creating this issue to track and join discussions in one place.

**Comment by benjamin-asdf** [comment]
Yes please! This is missing.

**Comment by ccleee** [comment]
Yes please, when use cuda graph which cannot use device/stream synchronize, I want to use stream call back to comfirm the sequence between gpu and some cpu oprations.
